### DISCUSSION TITLE
treesitter component

### DISCUSSION THREAD

**Discussion by sibouras** [rebelot/heirline.nvim#70] [Q&A] [answered]
hi im tryin to make a treesitter component that updates when `:TSToggle highlight` is triggered
so far this is what it looks like
```lua
local TreesitterActive = {
  condition = function()
    local ts_avail, ts = pcall(require, "nvim-treesitter.parsers")
    return ts_avail and ts.has_parser()
  end,

  -- update = { "BufWinEnter", "ModeChanged" },
  update = { "ModeChanged" },

  provider = function()
    -- local ts_avail, ts = pcall(require, "nvim-treesitter.parsers")
    -- return (ts_avail and ts.has_parser()) and "綠TS" or ""
    local buf = vim.api.nvim_get_current_buf()
    local highlighter = require("vim.treesitter.highlighter")
    return highlighter.active[buf] and "綠TS " or ""
  end,
  hl = { fg = "green", bold = false },
}
```
is this efficient?
does this line `local ts_avail, ts = pcall(require, "nvim-treesitter.parsers")`  run everytime a cursor changes?
i want the update table to have something like `LspAttach` `LspDetach` event but couldn't find a treesitter equivalent,
pls i need some feedback on this

**Comment by rebelot** [comment, marked as answer]
I think there's no need to check if TS is installed _each_ time the status line is refreshed. Instead, you could check it only once _outside_ the component. The valid update autocmd would be `OptionSet syntax`. Because conditions are evaluated regardless of a valid cache, you can also move the condition inside the provider to remove any operation on component evaluation.


```lua
local ts_avail, ts = pcall(require, "nvim-treesitter.parsers")

local TreesitterActive = {
  update = {
    "OptionSet",
    pattern = "syntax"
  }

  provider = function()
    if not (ts_avail and ts.has_parser()) then return end
    local buf = vim.api.nvim_get_current_buf()
    local highlighter = require("vim.treesitter.highlighter")
    return highlighter.active[buf] and "綠TS " or ""
  end,
  hl = { fg = "green", bold = false },
}
```

**Reply by sibouras** [reply]
> Because conditions are evaluated regardless of a valid cache, you can also move the condition inside the provider to remove any operation on component evaluation.

love how you can control everything in heirline, thank you for this amazing statusline 🤗

how would i add `BufWinEnter` event?
tried this `update = { "BufWinEnter", "OptionSet", pattern = "syntax" }`, it didn't work, tried other syntaxes got syntax errors, it works without the `pattern = "syntax"` but that would be inefficient

**Reply by rebelot** [reply]
Event, element [1] of update table, must be string or table of strings:

try
```
update = {{"BufWinEnter", "OptionSet}, pattern = "syntax"}
```

But idk if that would work, as pattern will be matched against buf name for the Bufwinenter event.. removing pattern should not decrease efficiency, as this component would be evaluated every time an option is set or you enter a new buffer, thus you can waste some fraction of millisecond on sporadic actions

**Reply by sibouras** [reply]
> update = {{"BufWinEnter", "OptionSet}, pattern = "syntax"}

tried this and got this error `All entries in 'event' must be strings`

> removing pattern should not decrease efficiency, as this component would be evaluated every time an option is set 

i removed pattern so with this `update = { "OptionSet" }`:  i put a print statement inside the provider function and it runs each time the statusline is refreshed, i expected it to be evaluated every time an option is set like you said but it runs everytime

is there another syntax i can try?

edit: i think `OptionSet` runs everytime other plugins keep using it like cmp, also when entering search mode or command mode the print statement runs

is that taxing? should i ignore it?

**Reply by rebelot** [reply]
I'm sorry, the correct syntax is

```
update = {
  "BufWinEnter",
  "OptionSet",
  pattern = "syntax"
}
```
Although the syntax is correct, it shouldn't work as the pattern makes no sense for a bufwinenter..

Try removing the pattern and use
os.clock to see if the message is actually changing

**Reply by sibouras** [reply]
> Although the syntax is correct, it shouldn't work as the pattern makes no sense for a bufwinenter..

yes bufwinenter got ignored
and yes the message changed, here's a demo of what happened

https://user-images.githubusercontent.com/64098483/194730970-6106da21-dee3-46f8-9119-93d3867629ad.mp4

**Reply by rebelot** [reply]
I understand. Try this:

```lua
local ts_avail, ts = pcall(require, "nvim-treesitter.parsers")
local TSHl = {
    update = {
        "OptionSet",
        pattern = "syntax",
    },
    init = function(self)
        if not rawget(self, 'once') then
            vim.api.nvim_create_autocmd('BufWinEnter', {
                callback = function()
                    self._win_cache = nil
                end
            })
            self.once = true
        end
    end,
    hl = function()
        local buf = vim.api.nvim_get_current_buf()
        local highlighter = require("vim.treesitter.highlighter")
        local fg = highlighter.active[buf] and 'green' or 'red'
        return { fg = fg, bold = true }
    end,
    provider = function()
        if not (ts_avail and ts.has_parser()) then
            return
        end
        return 'TS ' .. os.clock()
    end,
}
```

**Reply by sibouras** [reply]
this is perfect, thank you so so much <3
